### ISSUE TITLE
`help` command broken, `fish_config` opens in non-default browser

### ISSUE THREAD

**Issue by natesilva** [fish-shell/fish-shell#4035]
fish, version 2.5.0
Darwin MacBook-Pro-Nate.local 16.6.0 Darwin Kernel Version 16.6.0: Fri Apr 14 16:21:16 PDT 2017; root:xnu-3789.60.24~6/RELEASE_X86_64 x86_64
xterm-256color

When I type `help` I get the message:

```
0:77: execution error: "file:///usr/local/Cellar/fish/2.5.0/share/doc/fish/index.html" doesn’t understand the “open location” message. (-1708)
```

And when I type `fish_config`, the configuration opens, but in Firefox (which is not my default browser). In the terminal I see the messages:

```
Web config started at 'file:///Users/nate/.cache/fish/web_config-THIRE8.html'. Hit enter to stop.
0:69: execution error: "file:///Users/nate/.cache/fish/web_config-THIRE8.html" doesn’t understand the “open location” message. (-1708)
```

When I Googled for this, I found a similar-sounding issue with Jupyter Notebook that apparently is related to a Python change in macOS 10.12.5. Possibly related? https://github.com/jupyter/notebook/issues/2438

I upgraded to 10.12.5 yesterday — these commands were working before that.

**Comment by Konstruktionist** [comment]
I see the same problem, with a slightly different message for `fish_config`  
`Web config started at 'file:///Users/peter/.cache/fish/web_config-IFE4S3.html'. Hit enter to stop.
0:70: execution error: "file:///Users/peter/.cache/fish/web_config-IFE4S3.html" begrijpt het bericht 'open location' niet. (-1708)
69:77: execution error: application "chrome" kan niet worden opgevraagd.  (-1728)`

Also on macOS 10.12.5 & fish version 2.5.0

**Comment by krader1961** [comment]
I started seeing this a week or two ago. I'm on the macOS beta track so it is a certainty this problem was introduced by the macOS 10.12.5 update. I am labeling this issue an enhancement rather than bug because it appears the problem is not with fish itself.

**Comment by rodrigobdz** [comment]
I'm also having this problem.

1. I've tried setting the `BROWSER` environmental variable as suggested in the [python docs](https://docs.python.org/2/library/webbrowser.html) and also as shown in [stackoverflow](http://stackoverflow.com/a/24353812) with no success.

2. I also tried deleting Firefox. Calling `fish_config` opened a pop up window asking me to point out the location of Firefox. After I closed it and called `fish_config` again, I was redirected to Safari. Now that I've reinstalled Firefox, Safari is still being opened and the following error message is showing up since: 
`70:78: execution error: Can’t get application "firefox". (-1728)`

**Comment by faho** [comment]
How is this different from #2814?

**Comment by jacquesdong** [comment]
>> execution error: xxx  doesn’t understand the “open location” message. (-1708)

I've had this problem just after upgrade to macOS 10.12.5 yesterday,  and 

the "open location" error message seems to be raised by Apple Script, 

help and fish_config both use *osascript* to open GUI browser on macOS

## 1. help.fish
```fish
    # OS X /usr/bin/open swallows fragments (anchors), so use osascript
    # Eval is just a cheesy way of removing the hash escaping
    if test "$fish_browser" = osascript
        osascript -e 'open location "'(eval echo $page_url)'"'
        return
    end
```

## 2.  fish_config -> webbrowser.py
``` python
    class MacOSXOSAScript(BaseBrowser):
        def __init__(self, name):
            self._name = name

        def open(self, url, new=0, autoraise=True):
            if self._name == 'default':
                script = 'open location "%s"' % url.replace('"', '%22') # opens in default browser
            else:
                script = '''
                   tell application "%s"
                       activate
                       open location "%s"
                   end
                   '''%(self._name, url.replace('"', '%22'))
```

**Comment by krader1961** [comment]
This is definitely a bug in the macOS `osascript` command. If I type `osascript -i` to enter interactive mode then paste `open location "file:///usr/local/share/doc/fish/commands.html"` it tells me

```
!! "file:///usr/local/share/doc/fish/commands.html" doesn’t understand the “open location” message.
```

If I then paste the same command in the same osascript session it has the expected result. That command also works if run from the Script Editor app.

This is not a fish bug. Whether or not we can, or even should, work around this issue is TBD.

I did find that this works:

```
osascript -e 'tell application "Finder"'\n'open location "file:///usr/local/share/doc/fish/commands.html#string"'\n'end tell'
```

However the anchor is not passed thru to the browser so the user does not see the `string` documentation in that example. If, on the other hand I write `tell application "chrome"` (or `safari`) then the browser does get the anchor value.

**Comment by defjaf** [comment]
Yeah, this has come up in a few other places, too. It just manifested in Sierra 10.12.5. See http://www.andrewjaffe.net/blog/2017/05/python-bug-hunt.html.

**Comment by 100ideas** [comment]
Experiencing the same osascript bug - macOs 10.12.5, but setting the `$BROWSER` var to something other than path to Safari seems to fix it for me.

```bash
$ set -U BROWSER '/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome'

# chrome tab opens w/ expected content from 
# /usr/local/Cellar/fish/2.5.0/share/doc/fish/index.html
$ help    

# calls /usr/local/Cellar/fish/2.5.0/share/fish/tools/web_config/webconfig.py, which
# uses python webbrowser package, another victim of osascript bug(?) and the 
# source of the execution error message below (guessing).
#
# that said, another chrome tab magically opens with the configurator webapp, 
# which seems to function normally.
$ fish_config
Web config started at 'file:///Users/100ideas/.cache/fish/web_config-BNRYAT.html'. Hit enter to stop.
0:73: execution error: "file:///Users/100ideas/.cache/fish/web_config-BNRYAT.html" doesn’t understand the “open location” message. (-1708)
^C
Shutting down.
```

**Comment by zanchey** [comment]
If we can confirm the workaround, I will add something to the release notes for the next release.

**Comment by krader1961** [comment]
Explicitly setting `BROWSER` to the path of the browser program works for me on macOS Sierra.

**Comment by ridiculousfish** [comment]
I vaguely recall we used `osascript` instead of `open` because `open` didn't support URL #anchors but it appears that it does now. Can we simply replace the `osascript` call with `open`?

**Comment by ridiculousfish** [comment]
Testing a fix now

**Comment by ridiculousfish** [comment]
This seems to do it for me but of course I would appreciate some testing.

**Comment by zanchey** [comment]
At this stage, I think we should hold on releasing 2.6.0 until this is fixed - web configuration is one of our key features.

**Comment by kballard** [comment]
@ridiculousfish I just tested this on 10.12.5, and while it does fix the osascript issue, `open` definitely does not handle anchors, so `help` is still pretty useless.

**Comment by kballard** [comment]
I think I just found a workaround. Trying to open the URL twice from osascript works. We can use a try-catch to simply repeat the open command.

**Comment by kballard** [comment]
I'm not sure how to apply that to `fish_config`, because I don't have any experience with the `webbrowser` module, though looking at 2f5af733ba0d75e8d55960a6e4bfe63268c24b59 it appears that it's not actually trying to use an anchor at all, so it should probably just be changed to not test the OS version and simply always use `open` (/cc @ridiculousfish). It's only `help` that cares about anchors.

**Comment by krader1961** [comment]
That's weird. I could have sworn that when I tested the `open` workaround it worked handled anchors yet when I repeated the test just now it doesn't. Chalk it up to not paying close enough attention. And, yes, I too noticed that if I ran `osacript` interactively and did the open twice in a row the second one works. See my earlier [comment](https://github.com/fish-shell/fish-shell/issues/4035#issuecomment-302599316). Exactly how to do that non-interactively is not clear.

**Comment by kballard** [comment]
@krader1961 I have a fix I'll be submitting a PR for in a minute. But I'm only touching `help`, see above comment about `fish_config`.

**Comment by kballard** [comment]
Submitted as #4083. I didn't mark that PR as fixing this ticket because I think we should rip out the version check in fish_config before considering this done, but I'm not confident enough to do that myself.

**Comment by defjaf** [comment]
Apparently, the underlying `osascript` bug is [fixed in the latest macOS 10.12.6 beta](http://bugs.python.org/issue30392#msg294865), so that part of the problem will be fixed with the next macOS update (the problem with explicitly setting Chrome is separate, but won't be an issue if Chrome is set as the default browser).

**Comment by zanchey** [comment]
`fish_config` does sort of care about anchors - if you use `fish_config func` it opens the functions tab, for example.

**Comment by krader1961** [comment]
Closing since a) we've introduced changes to work around the broken behavior, and b) it appears that macOS itself has a fix for this problem.

**Comment by zanchey** [comment]
I'm still trying to work out how fixed this is for the release.

If it's going to be fixed in 10.12.6, I'm tempted to back out the `fish_config` changes - the web address to visit is printed to the terminal and can be copy-pasted, and it will only be broken in 10.12.5.

Thoughts?

**Comment by krader1961** [comment]
I don't have any objection to backing out that fix if you're so inclined. I've confirmed, after upgrading my server to 10.12.6 last night, that the osascript bug has been fixed.
